single_kernel_mongo/managers/sharding.py

```diff
@@ -0,0 +1,689 @@
+            logger.info("Waiting for secrets from config-server")
+            raise WaitingForSecretsError("Missing keyfile")
+
+        self.update_member_auth(keyfile, tls_ca)
+
+        if tls_ca is not None and self.dependent.tls_manager.is_waiting_for_both_certs():
+            logger.info("Waiting for requested certs before restarting and adding to cluster.")
+            raise WaitingForCertificatesError
+
+        if not self.dependent.mongo_manager.mongod_ready():
+            raise NotReadyError
+
+        if not self.charm.unit.is_leader():
+            return
+
+        operator_password = self.state.shard_state.operator_password
+        backup_password = self.state.shard_state.backup_password
+        if not operator_password or not backup_password:
+            raise WaitingForSecretsError("Missing operator password or backup password")
+
+        self.sync_cluster_passwords(operator_password, backup_password)
+
+        # We have updated our auth, config-server can add the shard.
+        self.data_requirer.update_relation_data(relation.id, {"auth-updated": "true"})
+        self.state.app_peer_data.mongos_hosts = self.state.shard_state.mongos_hosts
+
+    def handle_secret_changed(self, secret_label: str | None) -> None:
+        """Update operator and backup user passwords when rotation occurs.
+
+        Changes in secrets do not re-trigger a relation changed event, so it is necessary to listen
+        to secret changes events.
+        """
+        if not self.charm.unit.is_leader():
+            return
+        if not secret_label:
+            return
+        if not (relation := self.state.shard_relation):
+            return
+        if self.data_requirer.fetch_my_relation_field(relation.id, "auth-updated") != "true":
+            return
+
+        # many secret changed events occur, only listen to those related to our interface with the
+        # config-server
+        sharding_secretes_label = f"{self.relation_name}.{relation.id}.extra.secret"
+        if secret_label != sharding_secretes_label:
+            logger.info(
+                f"Secret unrelated to this sharding relation {relation.id} is changing, ignoring event."
+            )
+            return
+
+        operator_password = self.state.shard_state.operator_password
+        backup_password = self.state.shard_state.backup_password
+
+        if not operator_password or not backup_password:
+            raise WaitingForSecretsError("Missing operator password or backup password")
+        self.sync_cluster_passwords(operator_password, backup_password)
+
+    def drain_shard_from_cluster(self, relation: Relation) -> None:
+        """Waits for the shard to be fully drained from the cluster."""
+        self.assert_pass_hook_checks(relation, is_leaving=True)
+
+        self.charm.status_manager.to_maintenance("Draining shard from cluster.")
+
+        mongos_hosts = self.state.app_peer_data.mongos_hosts
+
+        self.wait_for_draining(mongos_hosts)
+
```

Comment:
is this whitespace on purpose?

Reply:
I don't see any whitespace in this strings?
I'll run the formatter again which removes whitespaces.